test/common/firefox-cdp-driver.js

```diff
@@ -373,37 +374,34 @@ CDP(options)
                         break;
                     let command = input_buf.slice(0, i);
 
-                    // initialize loadEventFired promise for every command except expectLoad() itself (as that
-                    // waits for a load event from the *previous* command); but if the previous command already
-                    // was an expectLoad(), reinitialize also, as there are sometimes two consecutive expectLoad()s
-                    if (!pageLoadPromise || !command.startsWith("expectLoad("))
-                        pageLoadPromise = new Promise((resolve, reject) => { pageLoadResolve = resolve; pageLoadReject = reject; });
-
                     // HACKS: See description of related functions
                     if (command.startsWith("client.Page.addScriptToEvaluateOnNewDocument"))
                         command = command.substring(12);
 
                     // run the command
+                    let seq = ++cur_cmd_seq;
                     eval(command).then(reply => {
+                        currentExecId = null;
                         if (unhandledExceptions.length === 0) {
-                            success(reply);
```

Comment:
So if I understood the  above "Firefox does not report an error when the execution context.. got destroyed" correctly: Could we instead move the hack here? Then the `currentExecId` hack would at least stay localized to this command loop.

Instead, there could be another global variable, a set `validContexts` which gets updated in `executionContext{Created,Destroyed}`, in addition to `frameIdToContextId`. Then we could check for the validity of `currentExecId` here?

This would avoid the `currentExecId` global variable, which only has a well-defined meaning under the assumption that only a single command runs at a time. This seems currently true with our cdp.py, but is still error prone IMHO -- e.g. it's not unthinkable that some execution context from an earlier operation gets destroyed later on, and then `currentExecId` would be wrong.

Reply:
> Could we instead move the hack here? 

No, I think that wont work.  The `eval.command()` promise will never settle, and the whole machinery just hangs. We need to send a reply out from the contextDestroyed event handler since that is the only place that will run.

Reply:
If we want to support multiple concurrent commands (ugh!), we can do that by having a global seq -> contextId mapping. However, if we support concurrent commands, we need some protocol anyway to associate replies with requests (seqno maintained by cdp.py, say), and we should hook into that with the context tracking. But before have that, I'd rather not code up something complicated that is not used and will very likely need to be replaced anyway.

Reply:
Sorry, I did not mean to suggest that we should support multiple concurrent commands. My concern is that we can't be sure that all pending contexts get destroyed before a command ends. It could be that a context destroy event happens that applies to the *previous* command, and then would fail the current one. That logic is really hard to validate.

Reply:
OK, thanks; I hope we'll be able to debug that in a few months :grin: